### PR TITLE
Show more information in the security profiler

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17856 |
| License | MIT |
| Doc PR | - |

This is an early prototype to explore the feature of displaying more information in the security panel. Example:

profiler_security

### CODE REVIEW

**Comment by HeahDude** [comment]
Really awesome ! Many thanks for that feature !

File: src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authorization;
+
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+/**
+ * AccessDecisionManager is the base class for all access decision managers
+ * that use decision voters.
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: This docblock looks wrong.

Comment (javiereguiluz) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: Fixed. Thanks.

**Comment by fabpot** [comment]
@javiereguiluz Can you rebase as tests have been fixed since your submitted the PR.

File: src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+/**
+ * Decorates the original AccessDecisionManager class to log information
+ * about the security voters and the decisions made by them.
+ *
+ * @author Javier Eguiluz <javier.eguiluz@gmail.com>
+ */
+class DebugAccessDecisionManager implements AccessDecisionManagerInterface //extends AccessDecisionManager
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: The comment should probably be removed here, right?

File: src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+     * @return string
+     */
+    public function getStrategy()
+    {
+        // The $strategy property stores the name of its method, instead of the
+        // original strategy name. Example:
+        // method name = 'decideAffirmative', strategy = 'affirmative'
+        $strategyName = strtolower(substr($this->strategy, 6));
+
+        return $strategyName;
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: You can probably return above without the intermediate variable.

File: src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+namespace Symfony\Component\Security\Core\Authorization;
+
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+/**
+ * Decorates the original AccessDecisionManager class to log information
+ * about the security voters and the decisions made by them.
+ *
+ * @author Javier Eguiluz <javier.eguiluz@gmail.com>
+ */
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: This class should probably be marked as being `@internal`.

**Comment by fabpot** [comment]
Apart from my minor comments, :+1:

**Comment by javiereguiluz** [comment]
I've made all the requested changes (thanks for the review!) ... but before merging this, it should be tested by someone which works on Symfony apps with complex security needs (maybe @iltar could help us?)

**Comment by iltar** [comment]
@javiereguiluz sadly still running on 2.8 in the application I can test it on properly. The 3.0 upgrade will be done somewhere starting in 2 weeks as we are still gathering deprecations in our production env. Maybe I'll be able to upgrade just the security bundle as most (if not all) of those are already fixed.

**Comment by iltar** [comment]
@javiereguiluz I've downloaded your branch as zip, extracted it and symlinked vendor/symfony/symfony to your branch. However, I cannot get it to work. The error I'm getting is `Warning: Invalid argument supplied for foreach()`:

### issue 1
Most important part of the stacktrace:
```
in SecurityDataCollector.php line 117
at ErrorHandler->handleError('2', 'Invalid argument supplied for foreach()', '/home/ivanderberg/projects/forks/symfony-fix_17856/src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php', '117', ...)
```

I've checked the service and it has the `DebugAccessDecisionManager` which is correct. However, `getVoters()` returns `null`. I've checked and `setVoters()` is never called. I've traced that back to `AddSecurityVotersPass:48`:
```php
$container->getDefinition('security.access.decision_manager')->addMethodCall('setVoters', array(array_values($voters)));
```

There's 2 possible solutions:
 - `DebugAccessDecisionManager` setVoters()` should be removed and `getVoters()` should forward the call to `$this->manager`. Sadly this class doesn't have `getVoters()` yet so it will have to be added.
 - The `setVoters()` should also be called on the decorating service. It seems like this is not added because the decoration pass is later (correct me if I'm wrong)

----

### issue 2
After patching either one of the above issues, I have a serialization issue "You cannot serialize or unserialize PDO instances". The @Template is being cached now and causes the bug I fixed in this pending PR: sensiolabs/SensioFrameworkExtraBundle#404. However, fixing this issue didn't solve my serialization issue. After digging a bit deeper it made sense... The request also contains your session. This means that if you use a database session storage like I do, it will try to serialize this. This will also fail if you use the `@ParamConverter` and put objects in your attribute bag.

I've changed `DebugAccessDecisionManager:50` to `'object' => is_object($object) ? get_class($object) : gettype($object),` and now it works. In the actual PR you might want to use the ClassUtils from doctrine to fetch the actual class though, otherwise you'll see useless proxy names.

----

### After fixing

The following screens are on my login page. The first thing I notice is that the Token column might be redundant.



----

When I'm logged in, it shows more information on my dashboard. The things I notice:
 - The attributes are not shown in the most readable format for the most common case (1 attribute).
 - Column numbers get put on multiple lines
 - The Token column is redundant
 - I'm unable to trace back where it came from (but this could be a future PR)




----
When viewing a page which as _a lot_ of `isGranted()` checks, it becomes hard to trace back the data. Maybe it could be grouped but that would be annoying to trace back either way. A nice feature that I'm missing is the ability to show the "identifier" of an entity in this case. In this list it looks like I have inconsistent results while in fact it's 80 different entities being checked in a list. Finding the identifier might be nice as it will help debugging a lot but you will run into the same issue I have: iltar/http-bundle#12.


_and the list goes on_


----
Looking at a page with less different objects, it becomes a bit of a mess as well.


----

### Conclusion
It's a nice start but with lots of information it becomes a mess really quick. A possible solution would be to group the checks per object (via object hash?) and to add the ID of an object if managed by doctrine (plug & play system so people can hook in?). Additionally the last column could be removed and the attributes shown with the just attribute as comma imploded string.

**Comment by javiereguiluz** [comment]
@iltar thank you very much for your incredible review!!

I've made some changes as requested:

 * Removed the `token` column in the decision log
 * Improved the way we display attributes when there is only 1 attribute
 * Tried to improve the way objects are represented (this needs to be tested)
 * Related to previous one: don't store objects to avoid serializing errors
 * Other minor design tweaks

I still need to fix the `setVoters()` problem.

**Comment by iltar** [comment]
Once the `setVoters()` is fixed I'll run it through the application again if I have some time today, otherwise it will be tomorrow. Can you ping me when it's ready to be tested?

**Comment by javiereguiluz** [comment]
@iltar it should be ready for review. Thanks!

**Comment by iltar** [comment]
Not sure if the Request gives the desired effect now, but works for me:


----
The list is a lot easier to read now but differences between objects is still hard to find. I suppose getting the actual ID field + the value of the object in question (if entity) would be too hard to implement in this patch for now. It's already an improvement but it takes a lot of time to see the different objects.

*2 different objects*


*Many different objects*


*Found one with a basic type, someone found it necessary to use an ID instead of an object...*
image

Besides of the minor comments (which are hard to fix), I think this shows a lot of usable information already. What I would still like to see is a way to identify an Entity in a generic fashion; I need it mysel and it could be re-used here to show useful info.

----

I hope this helps :+1:

**Comment by javiereguiluz** [comment]
@iltar great! We're making progress here. I've just added a special case to check if `getId()` method exists to use it instead of the object hash. I think it's worth it because `getId()` is used very commonly.

**Comment by fabpot** [comment]
Good to merge? ping @symfony/deciders

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php [hunk 5]
```diff
@@ -104,6 +108,20 @@ public function collect(Request $request, Response $response, \Exception $except
                 'supports_role_hierarchy' => null !== $this->roleHierarchy,
             );
         }
+
+        // collect voters and access decision manager information
+        if (null !== $this->accessDecisionManager) {
+            $this->data['access_decision_log'] = $this->accessDecisionManager->getDecisionLog();
+            $this->data['voter_strategy'] = $this->accessDecisionManager->getStrategy();
```
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Both the `getDecisionLog()` and `getStrategy()` methods may not exist.

Comment (iltar) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Could typehint against the `DebugAccessDecisionManager`, good to let it crash if you run this incorrectly imo

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: I've made this change. Is this what you suggested?

```diff
-if (null !== $this->accessDecisionManager) {
+if ($this->accessDecisionManager instance of DebugAccessDecisionManager) {
```

**Comment by javiereguiluz** [comment]
I've tested all kinds of attributes and objects and it seems to work as expected:

access_decision_log

The Travis errors are unrelated.

**Comment by iltar** [comment]
That looks really nice and useful! :+1:

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Comment by felipsmartins** [comment]
Great great great!!! Congrats @javiereguiluz @iltar

**Pull request closed without merging**